File: InformationTheory2015/Assignment01/problem2.py
import numpy as np
import random
import networkx as nx
import matplotlib.pyplot as plt

class Coin():
    def __init__(self, id):
        self.weight = 0
        self.id = id

def coins_with_one_counterfiet(n):
    coins = [ Coin(x) for x in range(n-1) ]
    # Lighter of heavier coin
    c = Coin(n)
    c.weight = random.choice([-1,1])
    coins.append(c)
    random.shuffle(coins)
    return coins

def weight(coins):
    return sum([x.weight for x in coins ])

def split_coins(coins):
    """
    Split the coins into two equal parts. 

    If the coins are odd numbers, then leave one coins. and check the rest.
    """
    extra = None
    if len(coins) % 2 == 0: 
        extra = None
        left, right = coins[0:len(coins)/2], coins[len(coins)/2:]
    else: 
        extra = coins[-1]
        left, right = coins[0:len(coins)/2], coins[len(coins)/2:-1]
    return left, right, extra

def draw_graph(graph, plot = True, outfile = None):
    nx.draw(graph)
    if plot:
        plt.show()
    elif outfile:
        print("[INFO] Saving files to %s" % outfile)
        plt.savefig(outfile)

def branch(graph, node_id):
    """Create a computational branch and return the nodes to execute the next
    step of algorithm
    """
    coins = graph.node[node_id]['coins']
    left, right, extra = split_coins(coins)
    if extra is None:
        graph.add_node(node_id + 1, coins = left)
        graph.add_edge(node_id, node_id + 1)
        graph.add_node(node_id + 2, coins = right)
        graph.add_edge(node_id, node_id + 2)

def solve(graph, node_id):
    """Takes a graph with first node as all coins. Recurse over it to solve the
    graph"""
    n = branch(graph, node_id)
    draw_graph(graph)


def main():
    coins = coins_with_one_counterfiet(12)
    coinGraph = nx.Graph()
    coinGraph.add_node(0, coins = coins, weight = weight(coins))
    solve(coinGraph, 0)

if __name__ == '__main__':
    main()

File: InformationTheory2015/Assignment01/problem2.hs
-- Problem 2
import System.Random
import Data.List

data LeftSideOfBalance = Light | Equal | Heavy deriving (Eq, Show, Ord)

-- args: number of coins, counterfiet position and its weight
-- Create a list of coins.
generate_coins :: Int -> Int -> Float -> [Float]
generate_coins ncoins pos w = helper ncoins pos w [] where 
    helper 0 pos w coins = coins
    helper n pos w coins 
        | n == pos = helper (n-1) pos w (w:coins)
        | otherwise = helper (n-1) pos w (1.0:coins)
--
-- This is our balance. We can only use this in our algorithm.
balance :: [Float] -> [Float] -> LeftSideOfBalance
balance left right 
    | sum left == sum right = Equal
    | sum left > sum right = Heavy
    | otherwise = Light

solve coins = first_partition (splitIntoTwo coins)

--first_partition :: (Maybe Float, ([Float], [Float])) -> [(a, b)]
first_partition (Just x, (left, right)) 
    | Equal == balance left right = (Equal, [x])
    | otherwise = first_step left right (balance left  right)

first_partition (Nothing, (left, right)) 
    | Equal == balance left right  =  (Equal, [])
    | otherwise = first_step left right (balance left right)

first_step left right l_state = case (l_state, next_part_equal $ splitIntoTwo left) of 
    (Heavy, False) ->(Heavy, left) 
    (Light, False) ->(Light, left) 
    (Heavy, True) -> (Light, right)
    (Light, True) -> (Heavy, right)

next_part_equal (Just x, (left, right)) = case balance left right of 
    Equal -> Equal == balance [x] [head left]
    otherwise -> Equal == balance left right

-- Split the coins into two part. If odd, return the first coin along with split
-- of rest of coins. If even, just split into two parts.
splitIntoTwo :: [Float ] -> (Maybe Float, ([Float], [Float]))
splitIntoTwo coins 
    | odd (length coins) = (Just $ head coins, splitIntoTwo' $ tail coins)
    | otherwise = (Nothing, splitIntoTwo' coins)

-- Split the given even coins into two. 
splitIntoTwo' :: [Float] -> ([Float], [Float])
splitIntoTwo' coins  = splitAt (div (length coins) 2) coins

{-
solve coins = filter_phony (splitIntoTwo coins) Equal

filter_phony (Just x) left right prev = case (prev, balance' x left right) of
-}
    
            
-- Randomly generate n coins with one counterfiet.
main = do
    let ncoins = 7
    g <- getStdGen
    let pos = (fst $ randomR (1, ncoins) g) :: Int
    let weight = (fst $ randomR (1.5, 0.5) g) :: Float
    let coins  = generate_coins ncoins pos weight
    print $  coins
    print $ solve coins
    {-
    let ans = solve coins 
    print ans
    -}
    putStrLn "Done"
